core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -186,67 +212,71 @@ public void downgradeEdgeToInode(Inode inode, LockMode mode) {
-    LockResource lock = mInodeLockManager.lockEdge(edge, mode);
-    return new EdgeEntry(lock, edge, mode);
+    Edge lastEdge = lastEdge();
+    LockResource newEdgeLock = mInodeLockManager.lockEdge(lastEdge, LockMode.READ);
+    LockResource inodeLock = mInodeLockManager.lockInode(inode, mode);
+    removeLastLock();
+    addEdgeLock(lastEdge, LockMode.READ, newEdgeLock);
+    addInodeLock(inode, mode, inodeLock);
   }
 
   /**
-   * Locks an inode and creates and entry without adding the entry to the lock list yet.
-   *
-   * @return the inode entry
-   */
-  private InodeEntry createInodeEntry(Inode inode, LockMode mode) {
-    LockResource lock = mInodeLockManager.lockInode(inode, mode);
-    return new InodeEntry(lock, inode, mode);
-  }
-
-  /**
-   * If the lock list is write locked, returns LockMode.WRITE. Otherwise, returns the given mode.
+   * If mode is read but the lock list is write locked, returns LockMode.WRITE.
    *
    * This helps us preserve the invariant that there is never a READ lock following a WRITE lock.
    *
    * @param mode a lock mode
    * @return the mode
    */
   private LockMode nextLockMode(LockMode mode) {
-    if (endsInWriteLock()) {
-      return LockMode.WRITE;
-    }
-    return mode;
+    return endsInWriteLock() ? LockMode.WRITE : mode;
   }
 
-  /**
-   * Adds an entry to mEntries.
-   *
-   * @param entry the entry to add
-   */
-  private void addEntry(Entry entry) {
-    if (!endsInWriteLock() && entry.getMode() == LockMode.READ) {
-      mFirstWriteLockIndex++;
+  private void addLock(LockResource lock, LockMode mode) {
+    if (!endsInWriteLock() && mode == LockMode.WRITE) {
+      mFirstWriteLockIndex = mLocks.size();
     }
-    mEntries.add(entry);
+    mLocks.add(lock);
+  }
+
+  private void addInodeLock(Inode inode, LockMode mode, LockResource lock) {
+    mInodes.add(inode);
+    mLastEdge = null;
+    addLock(lock, mode);
+  }
+
+  private void lockAndAddInode(Inode inode, LockMode mode) {
+    addInodeLock(inode, mode, mInodeLockManager.lockInode(inode, mode));
+  }
+
+  private void addEdgeLock(Edge edge, LockMode mode, LockResource lock) {
+    mLastEdge = edge;
+    addLock(lock, mode);
+  }
+
+  private void lockAndAddEdge(Edge edge, LockMode mode) {
+    addEdgeLock(edge, mode, mInodeLockManager.lockEdge(edge, mode));
   }
 
   /**
-   * Removes and unlocks the last entry from mEntries.
+   * Removes and unlocks the last lock.
    */
-  private void removeLastEntry() {
-    mEntries.remove(mEntries.size() - 1).getLock().close();
-    if (mFirstWriteLockIndex > mEntries.size()) {
-      mFirstWriteLockIndex = mEntries.size();
+  private void removeLastLock() {
+    mLocks.remove(mLocks.size() - 1).close();
+    if (mFirstWriteLockIndex >= mLocks.size()) {
+      mFirstWriteLockIndex = NO_WRITE_LOCK_INDEX;
+    }
+    if (mLastEdge != null) {
+      mLastEdge = null;
+    } else {
+      Inode last = mInodes.remove(mInodes.size() - 1);
```

Comment:
Why is this required? Don't we already unlock and remove earlier?

Could this fail because there is nothing to remove?

Reply:
in line https://github.com/Alluxio/alluxio/pull/9092/files#diff-74f13917d59a65b6dcbc4f09e382edd3R265, we only remove the lock, here, we remove the corresponding inode.